source/common/grpc/subscription_impl.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include "envoy/config/subscription.h"
+#include "envoy/event/dispatcher.h"
+
+#include "common/grpc/async_client_impl.h"
+
+#include "api/base.pb.h"
+
+namespace Envoy {
+namespace Grpc {
+
+template <class ResponseType, class ResourceType>
+class SubscriptionImpl : public Config::Subscription<ResponseType, ResourceType>,
+                         AsyncClientCallbacks<ResponseType> {
+public:
+  SubscriptionImpl(const envoy::api::v2::Node& node, Upstream::ClusterManager& cm,
+                   const std::string& remote_cluster_name, Event::Dispatcher& dispatcher,
+                   const google::protobuf::MethodDescriptor& service_method)
+      : SubscriptionImpl(node, new AsyncClientImpl<envoy::api::v2::DiscoveryRequest, ResponseType>(
+                                   cm, remote_cluster_name),
+                         dispatcher, service_method) {}
+
+  SubscriptionImpl(const envoy::api::v2::Node& node,
+                   AsyncClient<envoy::api::v2::DiscoveryRequest, ResponseType>* async_client,
+                   Event::Dispatcher& dispatcher,
+                   const google::protobuf::MethodDescriptor& service_method)
+      : async_client_(async_client), service_method_(service_method),
+        retry_timer_(dispatcher.createTimer([this]() -> void { establishNewStream(); })) {
+    request_.mutable_node()->CopyFrom(node);
+  }
+
+  ~SubscriptionImpl() {
+    if (stream_ != nullptr) {
```

Comment:
This isn't the case in the gRPC streaming client right now, it expects explicit release of resources via the mutual close of both sides or a reset. I think I will fix this to make it more suitable for RAII.